### PR TITLE
Added qh3 as a package

### PR DESCRIPTION
**Pull request by thewchan**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/qh3/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  skip: True  # [py<37]
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+
+    # Adding this here as otherwise it relies on vendor lsqpack
+    - pylsqpack
```
Comment (xhochy) [inline recipes/qh3/meta.yaml:22]: This should be in host

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: let me try that

**Comment by thewchan** [comment]
@xhochy no dice still 
```
  src/qh3/_vendor/pylsqpack/binding.c:4:10: fatal error: lsqpack.h: No such file or directory
      4 | #include "lsqpack.h"
```

File: recipes/qh3/meta.yaml (hunk 1 above)
Comment (xhochy) [reply recipes/qh3/meta.yaml:22]: The compiler needs to stay in build.

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: got it I'll move it back

Comment (xhochy) [reply recipes/qh3/meta.yaml:22]: According to https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fpylsqpack-0.3.18-py39hf860d4a_0.conda the package doesn't contain the file in question.

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: strange... I guess I'm stuck then

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: The file is in the original ls-qpack C library: https://github.com/search?q=repo%3Alitespeedtech%2Fls-qpack+lsqpack.h&type=code

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: @xhochy does this mean the C library would need to be brought onto conda first?

Comment (xhochy) [reply recipes/qh3/meta.yaml:22]: Yes, we will need that first.

Comment (xhochy) [reply recipes/qh3/meta.yaml:22]: Depends on what the build tool is. Is it using autotools?

Comment (thewchan) [reply recipes/qh3/meta.yaml:22]: I checked upstream, seems like CMake?

File: recipes/qh3/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{% set name = "qh3" %}
+{% set version = "0.14.0" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/jawah/qh3/archive/v{{ version }}.tar.gz
```
Comment (Ousret) [inline recipes/qh3/meta.yaml:9]: hello there,

try this instead `https://github.com/jawah/qh3/releases/download/v{{ version }}/qh3-{{ version }}.tar.gz`
will solve the lsqpack missing header.

regards,

Comment (Ousret) [reply recipes/qh3/meta.yaml:9]: with `{% set version = "0.15.1" %}` as the starting pts.

**Comment by thewchan** [comment]
Hi @xhochy I'm pivoting back to this now, it's asking for `libclang` and I've added it to the build env, but somehow it's saying it still can't find it?

File: recipes/qh3/conda_build_config.yaml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+c_stdlib_version:       # [osx and x86]
+  - '10.12'             # [osx and x86]
```
Comment (xhochy) [inline recipes/qh3/conda_build_config.yaml:2]:
```suggestion
  - '10.13'             # [osx and x86]
```
10.13 is now the new default/minimum

File: recipes/qh3/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+
+requirements:
+  build:
+    - {{ stdlib('c') }}
+    - {{ compiler('c') }}
+    - {{ compiler('rust') }}
+    - cmake
+    - libclang  # [linux]
```
Comment (xhochy) [inline recipes/qh3/meta.yaml:23]:
```suggestion
```
This is something on PyPI only to reduce the size of binary artifacts (mostly?). In all recipes I have seen it, I just patched it out as a requirement.

Comment (thewchan) [reply recipes/qh3/meta.yaml:23]: I'll try Clang instead then

Comment (thewchan) [reply recipes/qh3/meta.yaml:23]: Or maybe I need to use Clang as the C compiler and not gcc?

Comment (thewchan) [reply recipes/qh3/meta.yaml]: @xhochy It seems that it's during the rust/cargo build phase it needs libclang, but not sure why it's not detecting it

**Comment by thewchan** [comment]
@xhochy New OSX errors

```
/Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/CommonCrypto/CommonRandom.h:35:9: error: unknown type name 'CCCryptorStatus'

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=932092&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=2120
```

Also linux build still not working, same issue, can't find libclang. I might try switching from gcc to clang

File: recipes/qh3/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    - {{ stdlib('c') }}
+    - {{ compiler('c') }}
+    - {{ compiler('rust') }}
+    - cmake
+    - clang  # [linux]
+    - nasm  # [win]
+  host:
+    - python
+    - maturin >=1.2,<2.0
+    - pip
```
Comment (xhochy) [inline recipes/qh3/meta.yaml:28]:
```suggestion
    - pip
    - clangdev
```
This is probably what you need

Comment (thewchan) [reply recipes/qh3/meta.yaml:28]: got it, let me try

**Comment by xhochy** [comment]
> @xhochy New OSX errors
> 
> 
> 
> ```
> 
> /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/CommonCrypto/CommonRandom.h:35:9: error: unknown type name 'CCCryptorStatus'
> 
> 
> 
> /Users/runner/Miniforge3/conda-bld/qh3_1715437872253/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/aws-lc-sys-0.15.0/aws-lc/crypto/fipsmodule/rand/urandom.c:382:42: error: use of undeclared identifier 'kCCSuccess'
> 
> ```
> 
> 
> 
> Also linux build still not working, same issue, can't find libclang. I might try switching from gcc to clang 

I think you need openssl in host and set OPENSSL_DIR=$PREFIX

**Comment by thewchan** [comment]
@xhochy maybe it's confused now because there is both gcc and clang? Or is it asking for a c++ compiler

```
warning: aws-lc-sys@0.15.0: Compiler version doesn't include clang or GCC: "/home/conda/staged-recipes/build_artifacts/qh3_1715440102416/_build_env/bin/x86_64-conda-linux-gnu-cc" "--version"
  warning: aws-lc-sys@0.15.0: Failed to run: "c++" "--version"
```

**Comment by xhochy** [comment]
Yes, it needs a C++ compiler

**Comment by thewchan** [comment]
@xhochy it still has trouble finding a compiler, I wonder if I need to specify or change some env variables for Clang?

```
 warning: aws-lc-sys@0.15.0: Compiler version doesn't include clang or GCC: "/home/conda/staged-recipes/build_artifacts/qh3_1715440942569/_build_env/bin/x86_64-conda-linux-gnu-cc" "--version"
  warning: aws-lc-sys@0.15.0: Compiler version doesn't include clang or GCC: "/home/conda/staged-recipes/build_artifacts/qh3_1715440942569/_build_env/bin/x86_64-conda-linux-gnu-c++" "--version"

  error: failed to run custom build command for `aws-lc-sys v0.15.0`


```

**Comment by xhochy** [comment]
No. This should be buildable with gcc. Switching to clang might even bring more hassle in this case.

**Comment by thewchan** [comment]
It really is this `aws-lc-sys-0.15.0` rust package causing all the problems in both linux and osx, but seems to be for different reasons

**Comment by xhochy** [comment]
I read a bit through its code. It seems like adding cmake to build and setting AWS_LC_SYS_CMAKE_BUILDER=1 may lead to better errors.

**Comment by thewchan** [comment]
@xhochy I already have cmake in build but let me try that env var

**Comment by thewchan** [comment]
Here's the additional error messages: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=932139&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1149

**Comment by xhochy** [comment]
You need make in build

File: recipes/qh3/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    - {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+
+requirements:
+  build:
+    - {{ stdlib('c') }}
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('rust') }}\
+    - clangdev
```
Comment (xhochy) [inline recipes/qh3/meta.yaml:26]: This should only be in host

File: recipes/qh3/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    - export AWS_LC_SYS_CMAKE_BUILDER=1
+    - {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+
+requirements:
+  build:
+    - {{ stdlib('c') }}
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('rust') }}\
```
Comment (xhochy) [inline recipes/qh3/meta.yaml:25]:
```suggestion
    - {{ compiler('rust') }}
```

**Review by xhochy — CHANGES_REQUESTED** [review]

**Comment by thewchan** [comment]
@xhochy Thanks so much! Linux build is fixed. Now trying to figure out the OSX build, I found this possible related situation but not sure if it is addressable here: https://github.com/openssl/openssl/issues/16458

**Comment by xhochy** [comment]
I'm a bit backlogged and thus I keep updating this branch but miss the time to work on the fix. Hopefully, I get to it this time 😰

**Comment by xhochy** [comment]
It works locally with a newer macOS SDK. Thus I would try to bump it a bit here and see whether the error vanishes. If not, I'll investigate further.

**Review by xhochy — APPROVED** [review]

**Comment by thewchan** [comment]
thanks so much @xhochy for helping with this!

**Pull request merged**